src/integrationTest/java/uk/gov/hmcts/reform/iacaseapi/integration/AppealReferenceNumberGeneratorIntegrationTest.java

```diff
@@ -0,0 +1,143 @@
+            appealReferenceNumberGenerator
+                .getNextAppealReferenceNumberFor(AsylumAppealType.RP.toString());
+
+        assertThat(appealReferenceNumber.get(), is("RP/50001/2018"));
+    }
+
+    @Test
+    public void generates_protection_appeal_reference_number_when_there_are_no_previous_appeals_in_ccd() throws JsonProcessingException {
+
+        givenCcd.doesntHaveAnyExistingAppealCases();
+
+        Optional<String> appealReferenceNumber =
+            appealReferenceNumberGenerator
+                .getNextAppealReferenceNumberFor(AsylumAppealType.PA.toString());
+
+        assertThat(appealReferenceNumber.get(), is("PA/50001/2018"));
+
+    }
+
+    @Test
+    public void generates_protection_appeal_reference_number_from_ccd_case_list() throws JsonProcessingException {
+
+        givenCcd.returns(
+            someListOfCasesIncludingButPriorTo(
+                anAsylumCase().withAppealReference("PA/50019/2018")));
+
+        Optional<String> appealReferenceNumber =
+            appealReferenceNumberGenerator
+                .getNextAppealReferenceNumberFor(AsylumAppealType.PA.toString());
+
+        assertThat(appealReferenceNumber.get(), is("PA/50020/2018"));
+    }
+
+    @Test
+    public void generates_revocation_of_protection_appeal_reference_number_from_ccd_case_list() throws JsonProcessingException {
+
+        givenCcd.returns(
+            someListOfCasesIncludingButPriorTo(
+                anAsylumCase().withAppealReference("RP/50019/2018")));
+
+        Optional<String> appealReferenceNumber =
+            appealReferenceNumberGenerator
+                .getNextAppealReferenceNumberFor(AsylumAppealType.RP.toString());
+
+        assertThat(appealReferenceNumber.get(), is("RP/50020/2018"));
+    }
+
+    @Test
+    public void generates_appeal_reference_numbers_for_each_type_from_ccd_case_list() throws JsonProcessingException {
+
+        givenCcd.returns(
+            someListOfCasesIncludingButPriorTo(
+                anAsylumCase().withAppealReference("RP/50019/2018")));
+
+        Optional<String> protectionAppealReferenceNumber =
+            appealReferenceNumberGenerator
+                .getNextAppealReferenceNumberFor(AsylumAppealType.PA.toString());
+
+        assertThat(protectionAppealReferenceNumber.get(), is("PA/50020/2018"));
+
+        Optional<String> revocationAppealReferenceNumber =
+            appealReferenceNumberGenerator
+                .getNextAppealReferenceNumberFor(AsylumAppealType.RP.toString());
+
+        assertThat(revocationAppealReferenceNumber.get(), is("RP/50020/2018"));
+    }
+
+    @Test
+    public void returns_error_when_unable_to_initialize_reference_numbers() {
+
+        givenCcd.hadAnInternalError();
+
+        Optional<String> appealReferenceNumber =
+            appealReferenceNumberGenerator
+                .getNextAppealReferenceNumberFor(AsylumAppealType.RP.toString());
+
+        assertFalse(appealReferenceNumber.isPresent());
```

Comment:
should this assertion be that an error was actually thrown? or is it ok to continue the submission without the `appealReferenceNumber`?

Reply:
@Rob-Doherty  It isn't the responsibility of the `appealReferenceNumberGenerator` to decide if the appeal should proceed or not. That decision is made in a better place.

Reply:
👌